### PR TITLE
Bugfix for model evaluation with MagUnit parameters 

### LINKED ISSUE #13133: Model evaluation fails if any model parameter is a `MagUnit` type value
### Description

As discovered by @larrybradley in PR #13116, models will fail to evaluate when one of the parameters has a `MagUnit`.

A simplified reproducer is the following code:
```python
from astropy.modeling.models import Const1D
import astropy.units as u

unit = u.ABmag
c = -20.0 * unit
model = Const1D(c)

model(-23.0 * unit)
```

This should evaluate cleanly to `-20.0 * unit`. Instead one gets the following traceback:
```python
---------------------------------------------------------------------------
UnitTypeError                             Traceback (most recent call last)
Input In [1], in <cell line: 8>()
      5 c = -20.0 * unit
      6 model = Const1D(c)
----> 8 model(-23.0 * unit)

File ~/projects/astropy/astropy/modeling/core.py:397, in __call__(self, model_set_axis, with_bounding_box, fill_value, equivalencies, inputs_map, *inputs, **new_inputs)
    390 args = ('self',)
    391 kwargs = dict([('model_set_axis', None),
    392                ('with_bounding_box', False),
    393                ('fill_value', np.nan),
    394                ('equivalencies', None),
    395                ('inputs_map', None)])
--> 397 new_call = make_function_with_signature(
    398     __call__, args, kwargs, varargs='inputs', varkwargs='new_inputs')
    400 # The following makes it look like __call__
    401 # was defined in the class
    402 update_wrapper(new_call, cls)

File ~/projects/astropy/astropy/modeling/core.py:376, in _ModelMeta._handle_special_methods.<locals>.__call__(self, *inputs, **kwargs)
    374 def __call__(self, *inputs, **kwargs):
    375     """Evaluate this model on the supplied inputs."""
--> 376     return super(cls, self).__call__(*inputs, **kwargs)

File ~/projects/astropy/astropy/modeling/core.py:1077, in Model.__call__(self, *args, **kwargs)
   1074 fill_value = kwargs.pop('fill_value', np.nan)
   1076 # prepare for model evaluation (overridden in CompoundModel)
-> 1077 evaluate, inputs, broadcasted_shapes, kwargs = self._pre_evaluate(*args, **kwargs)
   1079 outputs = self._generic_evaluate(evaluate, inputs,
   1080                                  fill_value, with_bbox)
   1082 # post-process evaluation results (overridden in CompoundModel)

File ~/projects/astropy/astropy/modeling/core.py:936, in Model._pre_evaluate(self, *args, **kwargs)
    933 inputs, broadcasted_shapes = self.prepare_inputs(*args, **kwargs)
    935 # Setup actual model evaluation method
--> 936 parameters = self._param_sets(raw=True, units=True)
    938 def evaluate(_inputs):
    939     return self.evaluate(*chain(_inputs, parameters))

File ~/projects/astropy/astropy/modeling/core.py:2704, in Model._param_sets(self, raw, units)
   2702             unit = param.unit
   2703         if unit is not None:
-> 2704             value = Quantity(value, unit)
   2706     values.append(value)
   2708 if len(set(shapes)) != 1 or units:
   2709     # If the parameters are not all the same shape, converting to an
   2710     # array is going to produce an object array
   (...)
   2715     # arrays.  There's not much reason to do this over returning a list
   2716     # except for consistency

File ~/projects/astropy/astropy/units/quantity.py:522, in Quantity.__new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    519         cls = qcls
    521 value = value.view(cls)
--> 522 value._set_unit(value_unit)
    523 if unit is value_unit:
    524     return value

File ~/projects/astropy/astropy/units/quantity.py:764, in Quantity._set_unit(self, unit)
    762         unit = Unit(str(unit), parse_strict='silent')
    763         if not isinstance(unit, (UnitBase, StructuredUnit)):
--> 764             raise UnitTypeError(
    765                 "{} instances require normal units, not {} instances."
    766                 .format(type(self).__name__, type(unit)))
    768 self._unit = unit

UnitTypeError: Quantity instances require normal units, not <class 'astropy.units.function.logarithmic.MagUnit'> instances.
```

I believe the issue might lie in `astropy.modeling.core` with this call:
https://github.com/astropy/astropy/blob/675dc03e138d5c6a1cb6936a6b2c3211f39049d3/astropy/modeling/core.py#L2703-L2704

I think more sophisticated logic for handling turning parameters into quantity like values needs to be included here, or possibly a refactor of the [`._param_sets`](https://github.com/astropy/astropy/blob/675dc03e138d5c6a1cb6936a6b2c3211f39049d3/astropy/modeling/core.py#L2662) method in general. I would like some input from those with more familiarity with the intricacies of the `astropy.units` for assistance with how to improve this logic.

### System Details

```
macOS-10.15.7-x86_64-i386-64bit
Python 3.9.10 (main, Feb  4 2022, 14:54:08)
[Clang 12.0.0 (clang-1200.0.32.29)]
Numpy 1.22.3
pyerfa 2.0.0.1
astropy 5.1.dev901+g675dc03e1
Scipy 1.8.0
Matplotlib 3.5.1
```

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

This is a bug fix for #13133, which reports that models cannot be evaluated with `MagUnit` parameters. This has been solved by enabling general use of `MagUnit` but with protections in the `Parameter` type.

Namely, when `mag=True` is passed to a model parameter, this parameter can take on `MagUnit` values as needed. This has been done because `MagUnit` quantities do not support multiplication and division generally, so `MagUnit` units cannot be supported in general by models. Thus if a model can operate with `MagUnit` as a parameter, it should specify `mag=True` when specifying the parameter.

Fixes #13133

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/modeling/powerlaws.py
Comment (pllim) [inline astropy/modeling/powerlaws.py:593]: There is no check that `m_star` and `mag` are in the same magnitude *system*?

File: docs/changes/modeling/13158.bugfix.rst
Comment (pllim) [inline docs/changes/modeling/13158.bugfix.rst:1]:
```suggestion
Bugfix for using ``MagUnit`` units on model parameters.
```

File: astropy/modeling/powerlaws.py
Comment (WilliamJamieson) [reply astropy/modeling/powerlaws.py:593]: Shouldn't that fail or raise a unit error?

Comment (pllim) [reply astropy/modeling/powerlaws.py:593]: 🤷

**Comment by pllim** [comment]
I tentatively milestoned to v5.1 but feel free to move milestone to v5.1.1 if this doesn't make it in time.

File: astropy/modeling/powerlaws.py
Comment (WilliamJamieson) [reply astropy/modeling/powerlaws.py:593]:
```python
import astropy.units as u

a = 5 * u.ABmag
b = 3 * u.STmag

print(a - b)
```
returns
```python
2.0 mag(AB / ST
```

Comment (WilliamJamieson) [reply astropy/modeling/powerlaws.py:593]: @larrybradley do `m_star ` and `mag` need to be in the same magnitude system?

Comment (pllim) [reply astropy/modeling/powerlaws.py:593]: That hurts my head... 😆 I'll defer to @larrybradley and @mhvk .

Comment (larrybradley) [reply astropy/modeling/powerlaws.py:593]: One should not be able to mix different magnitude systems because they have different zero points (and in some cases meanings, e.g. Vega mag vs. AB mag).  I don't know what a unit of `mag(AB / ST)` even means.  I would have expected an error.

Comment (WilliamJamieson) [reply astropy/modeling/powerlaws.py:593]: Alright I'll add a check to ensure that `m_star` and `mag` have the same units

Comment (larrybradley) [reply astropy/modeling/powerlaws.py:593]: Note that `m_star` and `mag` must be in magnitude units.  In the current implementation, someone could enter them both in arbitrary units (e.g., Jy), which would not raise an error, but also does not make sense.

In my original PR supporting mag units, I wrote a helper method to deal with the input units:
* `m_star` and `mag` must have identical units and they must be magnitude units
* 
https://github.com/larrybradley/astropy/blob/schechter-mag/astropy/modeling/powerlaws.py#L590

Comment (mhvk) [reply astropy/modeling/powerlaws.py:593]: In principle, the magnitudes should evaluate correctly. E.g.,
```
m_star = -20 * u.ABmag
mag = -22.5 * u.ABmag
factor = (m_star - mag).to(u.dimensionless_unscaled)
factor
# <Quantity 0.1>
```
And of course this will fail if the magnitudes are not in the same system.

Might it be easiest to just convert the inputs to `Magnitude` for this case?

File: astropy/modeling/parameters.py
Comment (mhvk) [inline astropy/modeling/parameters.py:196]: Is it necessary to add a new argument? I would think that for most models, magnitudes will just fail miserably if passed in. I.e., perhaps the check in `_set_unit` is not necessary? (But no real idea about the modelling API, so feel free to ignore!!)

File: astropy/modeling/powerlaws.py
Comment (mhvk) [reply astropy/modeling/powerlaws.py:593]: p.s. Incompatible magnitude units like `AB` and `ST` would raise an error if one not just subtracts, but tries to convert to dimensionless.

File: astropy/modeling/parameters.py
Comment (WilliamJamieson) [reply astropy/modeling/parameters.py:196]: Models which involve a multiply/divide of a magnitude parameter value fail in a horribly unintuitive way; moreover, we cannot support `MagUnit` type units in general way. This sets it up so that `MagUnit` type units will not be allowed in a parameter unless the model designer specifically allows them. The goal here is to allow `MagUnit` units in appropriate parameters while preventing uninformative tracebacks.

File: astropy/modeling/powerlaws.py
Comment (WilliamJamieson) [reply astropy/modeling/powerlaws.py:593]: Hey this is taken care of by the `input_units` method.

File: astropy/modeling/parameters.py
Comment (mhvk) [reply astropy/modeling/parameters.py:196]: OK, if the failure is hopeless unintuitive, then something to guard the user seems well worth it.

File: astropy/modeling/powerlaws.py
Comment (larrybradley) [inline astropy/modeling/powerlaws.py:595]:
```suggestion
        factor_exp = (magnitude - m_star)
        if isinstance(factor_exp, Quantity):
            return factor_exp.to(dimensionless_unscaled)
        else:
            return 10 ** (-0.4 * factor_exp)
```

This seems easier to read/understand to me.

**Comment by pllim** [comment]
If this cannot make it in time for v5.1, please remilestone for v5.1.1. Thanks!

**Review by larrybradley — APPROVED** [review]
This looks good to me.  The only thing I will point out is that the `u.mag` unit is not actually a `MagUnit`.  It's a `astropy.units.function.mixin.RegularFunctionUnit`.  Do we need to treat `u.mag` like `u.ABmag`, etc?

**Comment by perrygreenfield** [comment]
@larrybradley this is beginning to remind me of one of my favorite CS quotes: 

> If you think C++ is not overly complicated, just what is a protected abstract virtual base pure virtual private destructor and when was the last time you needed one?

**Comment by mhvk** [comment]
It is slightly worse than that (and I can say that as I wrote the code...), `u.mag` is not the same as `u.mag()` - only the latter is a proper logarithmic unit. In practice, the two behave mostly the same, but one cannot get the auto-conversion to dimensionless from the plain `Quantity`:
```
In [3]: (2.5*u.mag()).to(u.one)
Out[3]: <Quantity 0.1>

In [4]: (2.5*u.mag).to(u.one)
UnitConversionError: 'mag' and '' (dimensionless) are not convertible
```

I should probably note that really this is not a trivial problem: magnitudes (and dB and dex) are rather strange units, which behave differently!

File: astropy/modeling/powerlaws.py
Comment (WilliamJamieson) [inline astropy/modeling/powerlaws.py:599]: > It is slightly worse than that (and I can say that as I wrote the code...), `u.mag` is not the same as `u.mag()` - only the latter is a proper logarithmic unit. In practice, the two behave mostly the same, but one cannot get the auto-conversion to dimensionless from the plain `Quantity`:
> 
> ```
> In [3]: (2.5*u.mag()).to(u.one)
> Out[3]: <Quantity 0.1>
> 
> In [4]: (2.5*u.mag).to(u.one)
> UnitConversionError: 'mag' and '' (dimensionless) are not convertible
> ```
> 
> I should probably note that really this is not a trivial problem: magnitudes (and dB and dex) are rather strange units, which behave differently!

@mhvk does this address the issue? Note that there is are now test cases in `test_model_quantities.py:test_models_evaluate_magunits` with both `u.mag` and `u.ABmag` units.

Comment (mhvk) [reply astropy/modeling/powerlaws.py:599]: I'm confused how this can work for `ABmag`?! A `Magnitude` would pass the `isinstance` check, but the unit would not match. What am I missing?!

I'd have thought it would be something like this:
```
if isinstance(factor_exp, Quantity):
    try:
        if not isinstance(factor_exp, Magnitude):
            factor_exp = Magnitude(factor_exp.to_value(u.mag))
        return factor_exp.to(dimensionless_unscaled)
    except UnitsError as err:
        raise UnitTypeError("the units of magnitude and m_star must be magnitudes with matching physical type.") from err
```

Comment (WilliamJamieson) [reply astropy/modeling/powerlaws.py:599]: It is operating on `magnitude - m_star`.  Lets look at what gets returned:
```python
import astropy.units as u

magnitude_1 = -23 * u.ABmag
magnitude_2 = -23 * u.mag
m_star_1 = -20 * u.ABmag
m_star_2 = -20 * u.mag

factor_exp_1 = magnitude_1 - m_star_1
factor_exp_2 = magnitude_2 - m_star_2

print(isinstance(factor_exp_1, u.Quantity))
print(isinstance(factor_exp_2, u.Quantity))

print(factor_exp_1.unit == u.mag)
print(factor_exp_2.unit == u.mag)
```

Returns:
```python
True
True
True
True
```

It turns out that the unit on the difference of two `ABmag` quantities is `mag`.

Comment (mhvk) [reply astropy/modeling/powerlaws.py:599]: Ah, yes, of course! In principle that will change with #12579, which is why I was confused. But I think really your statement should work after that too, so I better just fix it so that `u.mag() == u.mag`. Anyway, no concerns for *this* PR, then!

File: astropy/modeling/tests/test_models_quantities.py
Comment (pllim) [inline astropy/modeling/tests/test_models_quantities.py:749]: Nitpick: Does `pytest.mark.skipif` not work here?

File: astropy/modeling/tests/test_models_quantities.py
Comment (pllim) [inline astropy/modeling/tests/test_models_quantities.py:765]: What about `model(-23 * u.mag)`? Is that expected to pass or fail here?

File: astropy/modeling/tests/test_parameters.py
Comment (pllim) [inline astropy/modeling/tests/test_parameters.py:467]:
```suggestion
        with pytest.raises(ValueError, match="This parameter does not support the magnitude units such as mag(AB)"):
            param._set_unit(u.ABmag, True)
```

p.s. I would rather you not use `black` to get formatting like this:

```python
x = (
    some_expression
)
```

**Review by pllim — COMMENTED** [review]
I need to test this with Vegamag. Will report back.

File: astropy/modeling/tests/test_models_quantities.py
Comment (WilliamJamieson) [reply astropy/modeling/tests/test_models_quantities.py:749]: Not all of the models involved require `scipy` (only those in `SCIPY_MODELS`), I don't think the `skipif` marker can distinguish among all the test cases in `mag_models`.

Comment (WilliamJamieson) [reply astropy/modeling/tests/test_models_quantities.py:749]: This is how all the other tests are handled in `test_models_quantities.py`

Comment (pllim) [reply astropy/modeling/tests/test_models_quantities.py:749]: Ah, okay. Thanks for the clarification!

File: astropy/modeling/tests/test_parameters.py
Comment (pllim) [reply astropy/modeling/tests/test_parameters.py:467]: Oh sorry, you need to do this:

```python
match=r"This parameter does not support the magnitude units such as mag\(AB\)")
```

Comment (WilliamJamieson) [reply astropy/modeling/tests/test_parameters.py:467]: It does not like how the `mag(AB)` string is formatted. Replacing that with a wild card fixes it.

File: astropy/modeling/tests/test_models_quantities.py
Comment (pllim) [inline astropy/modeling/tests/test_models_quantities.py:768]: Theoretically this makes no sense but it does not throw an exception here.

```python
>>> from synphot import units as syn_units
>>> model = Schechter1D(phi_star=1.e-4 * u.Mpc ** -3), m_star=-20. * syn_units.VEGAMAG, alpha=-1.9)
>>> model(17 * syn_units.OBMAG)
<Quantity 1.92431285e+09 1 / Mpc3>
```

Do I have to fix how I define these custom magnitude systems downstream?

https://github.com/spacetelescope/synphot_refactor/blob/e51d7605349eebea0921c7d72da674b0af7d6db9/synphot/units.py#L52-L55

Comment (pllim) [reply astropy/modeling/tests/test_models_quantities.py:768]: p.s. Is -20 ABmag in the test case realistic? 😸

Comment (pllim) [reply astropy/modeling/tests/test_models_quantities.py:768]: p.p.s. I guess it might be if this is the absolute magnitude? 🤷

Comment (larrybradley) [reply astropy/modeling/tests/test_models_quantities.py:768]: That's necessary to remove magnitude differences due to distance.

Comment (pllim) [reply astropy/modeling/tests/test_models_quantities.py:768]: Thanks! So, I guess that is a downstream issue and should not block this PR. It does throw exception with STmag/ABmag.

Then there is also OBMAG itself is instrument dependent but that is a headache for another day...

Comment (pllim) [reply astropy/modeling/tests/test_models_quantities.py:768]: Maybe add a warning about using custom mag units in the doc?

File: astropy/modeling/parameters.py
Comment (jianyangli) [inline astropy/modeling/parameters.py:178]: I recently ran into this issue of using magnitude units in model parameters too, and glad to see you guys are already working on it.  Thank you!

I just wonder if the `self_value = Quantity(self.value, self.unit)` lines (73 & 104) in the arithmetic operator wrappers should also have `subok=True` added?  This would allow direct arithmetic ops of parameters in the case of magnitude units.

Sorry that I can't insert my comment into the right place...

Comment (WilliamJamieson) [reply astropy/modeling/parameters.py:178]: This is rather dangerous since `MagUnit` based "quantities" only truly support subtraction (though addition is theoretically possible the unit of the result is not the same as that of the input units). The bigger issue is that multiplication and division are not supported e.g.:
```python
import astropy.units as u

(3 * u.ABmag) + (2 * u.ABmag)

astropy.units.core.UnitTypeError: Cannot multiply function quantities which are not dimensionless with anything.
```

Comment (mhvk) [reply astropy/modeling/parameters.py:178]: I'm confused by the example: that should work:
```
In [2]: (3 * u.ABmag) + (2 * u.ABmag)
Out[2]: <Magnitude 5. mag(AB2)>
```
But perhaps you meant to write a `*` instead of a `+`?

Anyway, I think the main problem is the one you note, that addition will typically *not* quite be what one perhaps naively expected: as one sees above, the unit of the magnitude is now `AB**2` - indeed, it makes little physical sense to add two `Magnitudes` - what does it mean to sum the magnitudes of a `V=20` and a `V=21` star? It really only makes sense to add a magnitude difference (which is dimensionless) to a magnitude.

Though I guess that might still be useful. However, I think this would need a bit more thought, so my sense for this PR specifically is just stick with what it does now -- which is very nice! And then perhaps open a separate issue about how best to deal with magnitudes in modelling more generally; I keep finding them very tricky (and I wrote the code!), and it would be good to start from some concrete examples of what one would like to fit.

Comment (WilliamJamieson) [reply astropy/modeling/parameters.py:178]: > But perhaps you meant to write a `*` instead of a `+`?

Yes... this is what I get for copying out of my terminal without checking.

Comment (WilliamJamieson) [reply astropy/modeling/parameters.py:178]: Note that at least within modeling, no arithmetic is ever done directly with a`Parameter` type (within any model or fitter), they are normally cast to a `Quantity` or pure value (depending on if operations support units or not).

Comment (jianyangli) [reply astropy/modeling/parameters.py:178]: OK, I see the concerns.  Yes, that totally makes sense!  Thank you for pointing it out.

I definitely support the idea of opening a separate issue to discuss magnitudes (and other logarithmic units), in a more general sense than just in modeling.  This type of unit is so tricky.  I can think of at least one different way of magnitude addition that is practically useful, e.g., in calculating the total magnitude of a binary from its components:

```python
import astropy.units as u

mag_a = 3 * u.ABmag
mag_b = 4 * u.ABmag
mag_tot = (mag_a.to(u.AB) + mag_b.to(u.AB)).to(u.ABmag)
print(mag_tot)
```

Of course this is not to be discussed here.  You may close this conversation.  Thanks!

Comment (mhvk) [reply astropy/modeling/parameters.py:178]: @jianyangli - sounds good! As implemented, one cannot add magnitudes like you suggested, but I think it might be useful to have something that, effectively, does the operation on the underlying physical quantity. Anyway, for another issue!

**Comment by pllim** [comment]
No further comments and has one approval, so I went ahead and merge. Thanks!

**Pull request merged**
